pkg/cloud/scope/cluster.go

```diff
@@ -23,16 +23,21 @@ import (
 	awsclient "github.com/aws/aws-sdk-go/aws/client"
 	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/klog/v2/klogr"
 	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/throttle"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
```

Comment:
It seems that this PR is causing the AWS infrastructure provider to take a dependency on both the kubeadm control plane and bootstrap providers. We should be very careful with making assumptions that Clusters use either of those as they are **providers** (as opposed to core CAPI) and by design should be modular. Doing this would essentially be saying "AWS is now only compatible with the kubeadm control plane and bootstrap providers" which would cause users to be forced to use those implementations. 

It's easy to forget since the kubeadm providers are part of the CAPI repo and releases, and there hasn't been much happening in terms of other (public) control-plane/bootstrap providers so far but if we want to keep the core design principle that CAPI is modular in its bootstrap and control plane implementations we should be adding tight coupling between infra providers and other providers.

cc @vincepri @fabriziopandini @enxebre

Reply:
Good point @CecileRobertMichon, this is not the way to go about this.

I don't know if there were discussions around adding common fields to the bootstrap providers as contract. The port  that api server will run on might be a common field.

I don't think of a way to extract this info from somewhere else without some API changes.

We can continue discussing this in the relevant CAPI issue: https://github.com/kubernetes-sigs/cluster-api/issues/5517

Reply:
I agree, this default infra implementation shouldn't preclude a consumer from plugging in a different Control Plane provider implementation, that modularity is critical for adoption and enables consumers to experiment and to develop support for new scenarios. I'm curious doesn't this even break your `ekscontrolplanev1.AWSManagedControlPlane` workflow? At the very minimum this should check the Control Plane ref API group and kind for this assumption to be valid. Though for a more sustainable approach that applies generically we could discuss about having the API server port as part of the contract ([1] and [2]) and let it be input for the control plane CR spec or something in that line. I'm also curious on how does this relate to the LB proposal.

[1] https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/architecture/controllers/cluster.md#contracts

[2] https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/architecture/controllers/control-plane.md#contracts

Reply:
@enxebre The proposal for adding contract looks good to me for a sustainable approach but I am not sure of it's relation to LB proposal. @randomvariable can you comment here?

Reply:
I think we need to hold this and take this to upstream, we can't make the current apiserver port field as part of the contract, given where it is located in the `kubeadmConfig` spec. Let's take the discussion to https://github.com/kubernetes-sigs/cluster-api/issues/5517